ndb/src/google/cloud/ndb/model.py

```diff
@@ -697,6 +701,236 @@ def __pos__(self):
+        * ``B._validate()``
+        * ``B._to_base_type()``
+        * ``A._validate()``
+        * ``A._to_base_type()``
+
+        Args:
+            value (Any): The value to be converted / validated.
+
+        Returns:
+            Any: The transformed ``value``.
+        """
+        methods = self._find_methods("_validate", "_to_base_type")
+        call = self._apply_list(methods)
+        return call(value)
+
+    def _call_shallow_validation(self, value):
+        """Call the "initial" set of ``_validate()`` methods.
+
+        This is similar to :meth:`_call_to_base_type` except it only calls
+        those ``_validate()`` methods that can be called without needing to
+        call ``_to_base_type()``.
+
+        An example: suppose the class hierarchy is
+
+        .. code-block:: python
+
+            class A(Property):
+                def _validate(self, value):
+                    ...
+                def _to_base_type(self, value):
+                    ...
+
+            class B(A):
+                def _validate(self, value):
+                    ...
+                def _to_base_type(self, value):
+                    ...
+
+            class C(B):
+                def _validate(self, value):
+                    ...
+
+        The full list of methods (in order) called by
+        :meth:`_call_to_base_type` is:
+
+        * ``C._validate()``
+        * ``B._validate()``
+        * ``B._to_base_type()``
+        * ``A._validate()``
+        * ``A._to_base_type()``
+
+        whereas the full list of methods (in order) called here stops once
+        a ``_to_base_type`` method is encountered:
+
+        * ``C._validate()``
+        * ``B._validate()``
+
+        Args:
+            value (Any): The value to be converted / validated.
+
+        Returns:
+            Any: The transformed ``value``.
+        """
+        methods = []
+        for method in self._find_methods("_validate", "_to_base_type"):
+            # Stop if ``_to_base_type`` is encountered.
+            if method.__name__ != "_validate":
+                break
+            methods.append(method)
+
+        call = self._apply_list(methods)
+        return call(value)
+
+    @classmethod
+    def _find_methods(cls, *names, reverse=False):
+        """Compute a list of composable methods.
+
+        Because this is a common operation and the class hierarchy is
+        static, the outcome is cached (assuming that for a particular list
+        of names the reversed flag is either always on, or always off).
+
+        Args:
+            names (Tuple[str, ...]): One or more method names to look up on
+                the current class or base classes.
+            reverse (bool): Optional flag, default False; if True, the list is
+              reversed.
+
+        Returns:
+            List[Callable]: Class method objects.
+        """
+        # Get cache on current class / set cache if it doesn't exist.
+        cache = cls.__dict__.get("_find_methods_cache")
```

Comment:
I **really** hate what's going on here. This is a mutable cache that is specific to each `Property` subclass (of which there are many). The only way to emulate this behavior that is "obvious" is a metaclass.

It's really no fun because every time a unit test calls a method that invokes `_find_methods`, it has side effects for all other tests (see the pytest fixture I added to account for this).

Reply:
Could you define the cache on the base class, and then use the fully-qualified name of `cls` as the key?

Reply:
Done